### ISSUE TITLE
How to Initialize a struct_list with Members from Byte Arrays

### ISSUE THREAD

**Issue by bdkiran** [capnproto/capnproto-rust#385]
I have a type that contains a list of another type. In my code I'm retriving the data on disk in the form of `Vec<Vec<u8>>` and looking to serialize those bytes to a list of that type in cap n proto. Based on looking through previous issues here, I believe this is the correct way to do this but the list always contains 0 items in it. Not sure what I'm doing wrong and why `set_with_caveats` does not return an error if it is unable to set the message.

```Rust
...
  let mut builder = Builder::new_default();
  //let mut list_builder: struct_list::Builder<message::Owned> = builder.init_root();
  let mut list_builder = builder.init_root::<struct_list::Builder<custom_message::Owned>>();
  list_builder.reborrow();
  
  for (i, list_item) in list_data.iter().enumerate() {
      let message_reader = serialize::read_message(
          list_item.as_slice(),
          ReaderOptions::new()
      ).unwrap();
      let reader = message_reader.get_root::<custom_message::Reader>().unwrap();
      let index = u32::try_from(i).unwrap();
      info!("CREATING LIST ind {} {:?}", thing, reader.get_key().unwrap());
      {
          let es = list_builder.reborrow().set_with_caveats(index, reader);
          match es {
              Ok(_l) => {
                  info!("Set message")
              },
              Err(err) => {
                  error!("{}", err)
              }
          }
      }
  }
  info!("Size {:?}", list_builder.len());
  
  let reader_list = list_builder.into_reader();

  info!("Size {:?}", reader_list.len());
...
```

Output
```
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] CREATING LIST ind 0 [107, 101, 121]
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] Set thing
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] CREATING LIST ind 1 [107, 101, 121]
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] Set thing
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] Size 0
[2023-03-15T04:31:08Z INFO  lucidmq::cap_n_proto_helper] Size 0
```

As you see here the reader has a value, but the `list_builder` has a size of 0. Unsure if my builder was initialized incorrectly or something else.

Referenced these issues already but solution is still unclear.
https://github.com/capnproto/capnproto-rust/issues/100
https://github.com/capnproto/capnproto-rust/issues/116

**Comment by dwrensha** [comment]
The problem is that [`message::Builder::init_root()`](https://github.com/capnproto/capnproto-rust/blob/62e9cd56eb31739784531f3774aa8d2d8ba79f44/capnp/src/message.rs#L440-L444) was not intended to be called on list types. I suppose that maybe we need to add something like `message::Builder::initn_root(length)`, similar to [`any_pointer::Builder::initn_as()`](https://github.com/capnproto/capnproto-rust/blob/62e9cd56eb31739784531f3774aa8d2d8ba79f44/capnp/src/any_pointer.rs#L153-L155).

**Comment by dwrensha** [comment]
> Not sure what I'm doing wrong and why set_with_caveats does not return an error

There was a missing bounds check. I added it in https://github.com/capnproto/capnproto-rust/commit/36441ca90155f3ffde3bcc5b881631a780b6b23a.

Your `list_builder` is getting initialized to length 0. In https://github.com/capnproto/capnproto-rust/commit/1ce11c2e9df9c50301a17d208d6ec457b2bf8a5e I've added a `message::Builder::initn_root()` method to make it possible to initialize the list to a nonzero length.
